File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
while True:
    sentence = input()
    if sentence == '.':
        break

    stack = []
    for i in sentence:
        if not stack:
            if i == ']' or i == ')':
                stack.append(i)
                break

            if i == '[' or i =='(':
                stack.append(i)

        else:
            if i == '[' or i =='(':
                stack.append(i)

            if i == ')':

                if stack[-1] == '[':
                    break

                elif stack[-1] == '(':
                    stack.pop()

            elif i == ']':
                if stack[-1] == '(':
                    break

                elif stack[-1] == '[':
                    stack.pop()
    if stack:
        print('no')
    else:
        print('yes')

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <iostream>
#include <stack>
#include <string>

using namespace std;

int main() {

	string str;
	while (true) {
		stack<char>s;
		string ans = "yes";
		getline(cin, str);
		if (str == ".") {
			break;
		}
		for (char c : str) {
			if (c == '[' || c == '(') {
				s.push(c);
			}
			else if(c == ']') {
				if (s.empty() || s.top() != '[') {
					ans = "no";
					break;
				}
				s.pop();
			}
			else if (c == ')') {
				if (s.empty() || s.top() != '(') {
					ans = "no";
					break;
				}
				s.pop();
			}
		}
		if (!s.empty()) {
			ans = "no";
		}

		cout << ans << '\n';
	}

	return 0;
}